reference-stores/next/context/Cart.js

```diff
@@ -1,205 +1,281 @@
+        linesArray.push(lineTransformer(lineItemsFromCart.splice(index, 1)[0]));
       }
+    });
+    linesArray = [
+      ...linesArray,
+      ...lineItemsFromCart.map((line) => lineTransformer(line))
+    ];
+    setLineItems(linesArray);
+    if (checkoutUrl) {
+      setCartCheckoutUrl(checkoutUrl);
     }
-  }, [cartItems]);
+  };
 
-  useEffect(() => {
-    setCartVisible(false);
-  }, [router.asPath]);
-
-  const addItem = (payload) => {
-    const index = cartItems.findIndex((lineItem) => {
-      if (lineItem.variantId === payload.variantId) {
-        return (
-          JSON.stringify(payload.metafields) ===
-          JSON.stringify(lineItem.metafields)
-        );
+  const setOptimisticCart = ({ lines, checkoutUrl }) => {
+    setLineItems(lines);
+    if (checkoutUrl) {
+      setCartCheckoutUrl(checkoutUrl);
+    }
+  };
+
+  const setErrors = ({ errors, userErrors }) => {
+    setCartErrors([]);
+    const errorsArray = [];
+    if (errors) {
+      errors.forEach((error) => errorsArray.push(error.message));
+    }
+    if (userErrors) {
+      userErrors.forEach((error) => errorsArray.push(error.message));
+    }
+    setCartErrors(errorsArray);
+  };
+
+  const addItem = async (line) => {
+    const index = lineItems.findIndex((lineItem) => {
+      if (line.cartLineId) {
+        return line.cartLineId === lineItem.cartLineId;
+      } else {
+        return lineItem.variantId === line.variantId;
       }
-      return false;
     });
     if (index === -1) {
-      payload.id = `${payload.variantId}::${uuid()}`;
-      setCartItems([...cartItems, payload]);
+      const cartLineId = `placeholder-${uuid()}`;
+      const items = [
+        ...lineItems,
+        {
+          ...line,
+          cartLineId
+        }
+      ];
+      setOptimisticCart({ lines: items });
+      setCartVisible(true);
+      const { cart, userErrors, errors } = await cartClient.cartLinesAdd({
+        cartId: cartId,
+        lines: [
+          {
+            nacelleEntryId: line.nacelleEntryId,
+            quantity: line.quantity || 1
+          }
+        ]
+      });
+      if (cart) {
+        setCart({ lines: cart.lines, checkoutUrl: cart.checkoutUrl });
+      }
+      setErrors({ userErrors, errors });
     } else {
-      setCartItems(
-        cartItems.map((cartItem, itemIndex) => {
-          if (index === itemIndex) {
-            return {
-              ...cartItem,
-              quantity: cartItem.quantity + payload.quantity
-            };
+      const items = [...lineItems];
+      const quantity = lineItems[index].quantity + line.quantity;
+      items.splice(index, 1, {
+        ...lineItems[index],
+        quantity
+      });
+      setOptimisticCart({ lines: items });
+      setCartVisible(true);
+      const { cart, userErrors, errors } = await cartClient.cartLinesAdd({
+        cartId: cartId,
+        lines: [
+          {
+            nacelleEntryId: line.nacelleEntryId,
+            quantity: line.quantity
```

Comment:
Should this be `quantity: quantity` instead of line.quantity? Can't tell from the cartLinesAdd mutation what happens with the quantity.

Reply:
Hey @mdarrik I had it at quantity originally but what it did is add previous quantity + new quantity to the added item, by using only the quantity from the payload it adds the correct amount

Reply:
Cool! I wasn't sure what would happen since Shopify doesn't really document their add line items mutation with that info. We should probably document that somewhere since it might be something merchant devs will run into.

Reply:
Yeah this caught me off-guard too! And definitely